tests/unit/test_utils.py

```diff
@@ -2316,20 +2332,65 @@ def test_disabling_env_var_not_true(self):
         url = 'https://example.com/'
         env = {'AWS_EC2_METADATA_DISABLED': 'false'}
 
-        self.add_get_token_imds_response(token='token')
-        self.add_get_role_name_imds_response()
-        self.add_get_credentials_imds_response()
+        self.add_default_imds_responses()
 
         fetcher = InstanceMetadataFetcher(base_url=url, env=env)
         result = fetcher.retrieve_iam_role_credentials()
 
         self.assertEqual(result, self._expected_creds)
 
+    def test_IMDS_USE_IPV6(self):
+        configs = [({'imds_use_ipv6': 'true'},'http://[fe80:ec2::254%eth0]/'),
+                ({'imds_use_ipv6': 'tRuE'}, 'http://[fe80:ec2::254%eth0]/'), 
+                ({'imds_use_ipv6': 'false'}, 'http://169.254.169.254/'), 
+                ({'imds_use_ipv6': 'foo'}, 'http://169.254.169.254/')]
+        for config, expected_url in configs:
+            yield _test_imds_base_url, config, expected_url
+    
+    def test_metadata_endpoint(self):
+        urls = ['http://fe80:ec2:0000:0000:0000:0000:0000:0000/', 'http://[fe80:ec2::010%eth0]/',
+                'http://192.168.1.1/']
+        for url in urls:
+            self.assertTrue(is_valid_ipv4_or_ipv6_endpoint(url))
+        
+    def test_IPv6_endpoint_no_brackets_env_var_set(self):
+        url = 'http://fe80:ec2::010/'
+        config = {'ec2_metadata_service_endpoint': url}
+        self.assertFalse(is_valid_IPv6_endpoint_url(url))
+
+    def test_IPv6_invalid_endpoint(self):
+        url = 'not.a:valid:dom@in'
+        config = {'ec2_metadata_service_endpoint': url}
+        with self.assertRaises(InvalidIMDSEndpointError):
+            InstanceMetadataFetcher(config=config)
+
+    def test_IPv6_endpoint_env_var_set_and_args(self):
+        url = 'http://[fe80:ec2::254]/'
+        url_arg = 'http://fe80:ec2:0000:0000:0000:8a2e:0370:7334/'
+        config = {'ec2_metadata_service_endpoint': url}
+
+        self.add_default_imds_responses()
+
+        fetcher = InstanceMetadataFetcher(config=config, base_url=url_arg)
+        result = fetcher.retrieve_iam_role_credentials()
+
+        self.assertEqual(result, self._expected_creds)
+        self.assertEqual(fetcher.get_base_url(), url_arg)
+
+    def test_IPv6_IMDS_not_allocated(self):
```

Comment:
It looks like you don't have any tests around supplying `None` or an empty config. Let's make sure we get those in there.